### PR TITLE
Improve rendering of incomplete evolution graph data points

### PR DESCRIPTION
**Pull request by mneudert**
### Current Evolution Graph Rendering

Evolution graphs display the line leading up the to last point as dashed if the last date/period contains today. For example when viewing the visitor overview for the current year (2024):

The visual change is only done for the very last point of the graph and can lead to unexpected results. Viewing the comparison of "current week" and "previous week" (current date is around 2024-01-10) displays multiple future dates but only one is incomplete:

### Suggested Evolution Graph Rendering

The PR changes the visual representation of incomplete data points to be a "circle" instead of a "dot" for better visibility, the dashed line leading up to an incomplete data point is also displayed. Additional support is added to represent any point in the graph as incomplete, including an improved detection of future dates that may get displayed (current date is around 2024-01-10):

To prepare for an upcoming improvement the state "invalidated" has already been included. If a report contains to-be-reprocessed (invalidated) information it will display this information in the graph for the specific invalidation dates, currently only available in tests:

Refs DEV-17556
Prepares DEV-17568

Moving the incomplete period detection to the backend will be a partial fix for #21560. The period selector will still be broken, but all graphs should be displayed as expected.

### Review

* [ ] [Functional review done](https://developer.matomo.org/guides/pull-request-reviews#functional-review-done)
* [ ] [Potential edge cases thought about](https://developer.matomo.org/guides/pull-request-reviews#potential-edge-cases-thought-about) (behavior of the code with strange input, with strange internal state or possible interactions with other Matomo subsystems)
* [ ] [Usability review done](https://developer.matomo.org/guides/pull-request-reviews#usability-review-done) (is anything maybe unclear or think about anything that would cause people to reach out to support)
* [ ] [Security review done](https://developer.matomo.org/guides/security-in-piwik#checklist)
* [ ] [Wording review done](https://developer.matomo.org/guides/pull-request-reviews#translations-wording-review-done)
* [ ] [Code review done](https://developer.matomo.org/guides/pull-request-reviews#code-review-done)
* [ ] [Tests were added if useful/possible](https://developer.matomo.org/guides/pull-request-reviews#tests-were-added-if-usefulpossible)
* [ ] [Reviewed for breaking changes](https://developer.matomo.org/guides/pull-request-reviews#reviewed-for-breaking-changes)
* [ ] [Developer changelog updated if needed](https://developer.matomo.org/guides/pull-request-reviews#developer-changelog-updated-if-needed)
* [ ] [Documentation added if needed](https://developer.matomo.org/guides/pull-request-reviews#documentation-added-if-needed)
* [ ] Existing documentation updated if needed

### CODE REVIEW

File: plugins/CoreVisualizations/JqplotDataGenerator/Evolution.php [hunk 1]
```diff
@@ -139,6 +139,18 @@ protected function initChartObjectData($dataTable, $visualization)
             }
             $visualization->setAxisXOnClick($axisXOnClick);
         }
+
+        $incompleteDataPoints = [];
+
+        foreach (array_values($dataTables) as $index => $dataTable) {
+            if (true !== $dataTable->getMetadata(DataTable::INCOMPLETE_ARCHIVE_METADATA_NAME)) {
```
Comment (mneudert) [inline plugins/CoreVisualizations/JqplotDataGenerator/Evolution.php]: The metadata key is only intended as a temporary solution to create a graph with multiple incomplete points.

Should later (i.e. DEV-17568) be replaced with a detection based on the used archives.

File: plugins/CoreVisualizations/javascripts/jqplot.js [hunk 2]
```diff
@@ -95,6 +95,55 @@ function rowEvolutionGetMetricNameFromRow(tr)
...
         },
 
+        _setIncompleteDataPoints: function (incompleteDataPoints) {
+            this.jqplotParams.incompleteDataPoints = [];
+
+            if (Array.isArray(incompleteDataPoints)) {
+                this.jqplotParams.incompleteDataPoints = incompleteDataPoints;
+            }
+
+            let tickCount = 0;
```
Comment (mneudert) [inline plugins/CoreVisualizations/javascripts/jqplot.js]: Once we have the backend detect incomplete data points, everything from here on in this method should be removed. The backend should already provide the information that the "today" data point is incomplete.

File: plugins/CoreVisualizations/javascripts/jqplot.js [hunk 3]
```diff
@@ -95,6 +95,55 @@ function rowEvolutionGetMetricNameFromRow(tr)
...
+                // If date is actually a range then adjust the period type for the containsToday check
+                period = 'range';
+            }
+
+            try {
+                if (piwikPeriods.parse(period, this.param.date).containsToday()) {
+                    this.jqplotParams.incompleteDataPoints.push(tickCount - 1);
+                }
+            } catch (e) {
+                // ignore period parsing error
```
Comment (mneudert) [inline plugins/CoreVisualizations/javascripts/jqplot.js]: Having the period parsing moved to a separate try-catch partially solves #21560, the graphs linked there should now always be rendered.

File: plugins/CoreVisualizations/javascripts/jqplotEvolutionGraph.js [hunk 4]
```diff
@@ -63,124 +61,125 @@
...
-                            }
-                        }
+                        content += `
+                            <h3 class="evolution-tooltip-header">${piwikHelper.htmlEntities(label)}</h3>
+                            ${dataByAxis[axisName].join('<br />')}
+                        `;
+                    }
+
+                    if (self.jqplotParams.incompleteDataPoints.includes(tick)) {
+                        content += `<br />(${self._lang.incompletePeriod})`;
```
Comment (mneudert) [inline plugins/CoreVisualizations/javascripts/jqplotEvolutionGraph.js]: Wording will most likely be updated in a separate PR (i.e. DEV-17569). Depending on the information available for incomplete data points (partially archived, not yet archived, invalidated, ...) we can show distinct variants.

File: tests/UI/specs/EvolutionGraphIncompleteData_spec.js [hunk 5]
```diff
@@ -0,0 +1,52 @@
...
+ * @license http://www.gnu.org/licenses/gpl-3.0.html GPL v3 or later
+ */
+
+describe('EvolutionGraphIncompleteData', function () {
+    const url = '?module=Widgetize&action=iframe&idSite=1&period=day&date=2012-01-31&evolution_day_last_n=30'
+              + '&moduleToWidgetize=UserCountry&actionToWidgetize=getCountry&viewDataTable=graphEvolution'
+              + '&isFooterExpandedInDashboard=1';
+
+    before(() => {
+        testEnvironment.forceIncompleteDates = [
```
Comment (mneudert) [inline tests/UI/specs/EvolutionGraphIncompleteData_spec.js]: Once the backend can provide this information, this variable (and the DI hook) can ideally be replaced with calls to archive invalidation.

File: tests/UI/expected-screenshots/Comparison_dashboard_last_period.png
Comment (bx80) [inline tests/UI/expected-screenshots/Comparison_dashboard_last_period.png:1]: Most of these screenshots seem to be identical, except in some the captured area is taller - are we expecting them to have changed? Are there some differences I'm missing?

Comment (sgiehl) [reply tests/UI/expected-screenshots/Comparison_dashboard_last_period.png:1]: That could be related to the menu being taller and should hopefully be fixed at some point with #21689

Comment (mneudert) [reply tests/UI/expected-screenshots/Comparison_dashboard_last_period.png:1]: This was the run I grabbed the screenshots from: https://github.com/matomo-org/matomo/actions/runs/7253959226/job/19761659351#step:3:1281

Only minimal changes for some line shadows (no idea where the height comes from, but that was also not a problem for the pixel diff detection :shrug:): https://builds-artifacts.matomo.org/matomo-org/matomo/dev-17556/7253959226/Comparison_dashboard_last_period.png

However those seem to stem from me messing up the `move to first non-null point` logic. Keeping it in place removed those unnecessary changes :see_no_evil:

File: plugins/CoreVisualizations/JqplotDataGenerator/Chart.php [hunk 6]
```diff
@@ -145,13 +150,14 @@ public function render()
...
-                'series' => &$this->series
-            ),
-            'data'   => &$this->data
-        );
+        $data = [
+            'params' => [
+                'axes' => &$this->axes,
+                'series' => &$this->series,
+            ],
+            'data' => &$this->data,
```
Comment (caddoo) [inline plugins/CoreVisualizations/JqplotDataGenerator/Chart.php:169]: Maybe a comment that the data variable should have the same amount of items as archive states and they are 1-1

File: plugins/CoreVisualizations/JqplotDataGenerator/Chart.php [hunk 7]
```diff
@@ -171,6 +177,14 @@ public function setAxisXLabelsMultiple($xLabels, $seriesToXAxis, $ticks = null)
         }
     }
 
+    /**
+     * @param array<string> $archiveStates
+     */
+    public function setArchiveStates(array $archiveStates): void
+    {
+        $this->archiveStates = $archiveStates;
```
Comment (caddoo) [inline plugins/CoreVisualizations/JqplotDataGenerator/Chart.php]: Related to comment about making it clear these are directly linked to data points, does it make sense to make this defensive and prevent someone passing 4 archive states for 3 data points for example?

**Review by caddoo — APPROVED** [review]
Went through it in a call with you, happy with the change just one small comment to make it clearer/more defensive for the next person who uses `setArchiveStates`

File: plugins/CoreVisualizations/JqplotDataGenerator/Chart.php (hunk 6 above)
Comment (mneudert) [reply plugins/CoreVisualizations/JqplotDataGenerator/Chart.php:169]: Renamed `archiveStates` to `dataStates` to be clearer what is meant.

File: plugins/CoreVisualizations/JqplotDataGenerator/Chart.php (hunk 7 above)
Comment (mneudert) [reply plugins/CoreVisualizations/JqplotDataGenerator/Chart.php]: Added a "consistency" check to log if the series or states have differing amounts of entries.

**Review by caddoo — APPROVED** [review]

File: plugins/CoreVisualizations/JqplotDataGenerator/Chart.php [hunk 8]
```diff
@@ -179,4 +208,28 @@ private function getXAxis($index)
         }
         return $axisName;
     }
+
+    private function checkDataAmountConsistency(): void
+    {
+        if ([] === $this->data && [] === $this->dataStates) {
```
Comment (michalkleiner) [inline plugins/CoreVisualizations/JqplotDataGenerator/Chart.php]: **Question** — I don't mind this way of checking for empty arrays, is that the way/standard we want to adopt in general? Otherwise I'd just go `empty($this->data)`. The strict check is better, but I suspect we may get a future type warnings as we are not probably in the state of code/data where we can reliably do this?

File: plugins/CoreVisualizations/javascripts/jqplot.js [hunk 9]
```diff
@@ -1658,56 +1677,74 @@ RowEvolutionSeriesToggle.prototype.beforeReplot = function () {
...
-                }
+        if (!points || !points.length) {
+            return;
+        }
+
+        let move = true;
+
+        for (let i = 0; i < points.length; i++) {
+            // skip to the first non-null point and move to it.
+            if (points[i][0] === null && points[i][1] === null) {
```
Comment (michalkleiner) [inline plugins/CoreVisualizations/javascripts/jqplot.js]: Style consistency Yoda for?

```suggestion
            if (null === points[i][0] && null === points[i][1]) {
```

**Review by michalkleiner — APPROVED** [review]
Code looks ok, just left one minor comment and one question.

File: plugins/CoreVisualizations/JqplotDataGenerator/Chart.php (hunk 7 above)
Comment (mneudert) [reply plugins/CoreVisualizations/JqplotDataGenerator/Chart.php]: Toned down the logs quite a bit. I missed the expected state that, when for example comparison "November" with "October", it is expected to have different amount of series points.

So only logging an info message if the state information not span the full graph ("longest series" should match state information count).

File: plugins/CoreVisualizations/JqplotDataGenerator/Chart.php (hunk 8 above)
Comment (mneudert) [reply plugins/CoreVisualizations/JqplotDataGenerator/Chart.php]: The default values are set to empty arrays and `setAxisYValues` is calling `$this->data[] = ...`. I would expect that to be as reliable as it currently can be :thinking:

Comment (michalkleiner) [reply plugins/CoreVisualizations/JqplotDataGenerator/Chart.php]: It was more a question around the style, it might be reliable here but if we adopt this approach more widely in other places it can be less reliable than `empty($array)`. We can discuss this outside of the PR, all good, was just curious.
